File: chroma-league/src/main/java/com/bonepl/chromaleague/tasks/MainThreads.java
package com.bonepl.chromaleague.tasks;

import com.bonepl.chromaleague.state.RunningState;
import com.bonepl.razersdk.ChromaRestSDK;

import java.io.Closeable;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;

public class MainThreads implements Closeable {
    public static final long PLAYER_LIST_FETCH_DELAY = 500L;
    public static final long ACTIVE_PLAYER_FETCH_DELAY = 100L;
    public static final long EVENTS_FETCH_DELAY = 300L;
    public static final long MAIN_HUD_REFRESH_DELAY = 50L;

    private static final Logger LOGGER = Logger.getLogger(MainThreads.class.getName());

    private final ScheduledExecutorService mainExecutor;
    private final GameLoader gameLoader;
    private final ChromaRestSDK chromaRestSDK;
    private boolean alive;

    public MainThreads() {
        alive = true;
        LOGGER.info("Game is loading");
        chromaRestSDK = new ChromaRestSDK();
        gameLoader = new GameLoader(chromaRestSDK);
        mainExecutor = Executors.newScheduledThreadPool(10);
        mainExecutor.scheduleWithFixedDelay(new FetchNewEventsTask(), 0L, EVENTS_FETCH_DELAY, TimeUnit.MILLISECONDS);
    }

    public void initializeGameThreads() {
        LOGGER.info("Player joined the game");
        gameLoader.close();
        RunningState.getGameState().setPlayerName(FetchPlayerName.fetchPlayerName());
        mainExecutor.scheduleWithFixedDelay(new FetchPlayerListTask(), 50L, PLAYER_LIST_FETCH_DELAY, TimeUnit.MILLISECONDS);
        mainExecutor.scheduleWithFixedDelay(new FetchActivePlayerTask(), 50L, ACTIVE_PLAYER_FETCH_DELAY, TimeUnit.MILLISECONDS);
        mainExecutor.scheduleWithFixedDelay(new RefreshMainHudTask(chromaRestSDK), 150L, MAIN_HUD_REFRESH_DELAY, TimeUnit.MILLISECONDS);
    }

    @Override
    public void close() {
        alive = false;
        RunningState.setRunningGame(false);
        shutdownMainExecutor();
        shutdownChromaSDK();
        LOGGER.info("Player left the game, waiting for another...");
    }

    private void shutdownMainExecutor() {
        try {
            mainExecutor.shutdown();
            if (!mainExecutor.awaitTermination(5000L, TimeUnit.MILLISECONDS)) {
                LOGGER.log(Level.WARNING, () -> "Couldn't terminate GameLoader executor - timed out");
            }
        } catch (InterruptedException e) {
            LOGGER.log(Level.WARNING, e, () -> "MainThreads interrupted while shutting down scheduler");
            Thread.currentThread().interrupt();
        } catch (Exception ex) {
            LOGGER.log(Level.WARNING, ex, () -> "Exception while shutting down main executor");
        }
    }

    private void shutdownChromaSDK() {
        try {
            chromaRestSDK.close();
        } catch (Exception ex) {
            LOGGER.log(Level.WARNING, ex, () -> "Exception while shutting down Chroma SDK");
        }
    }

    public boolean isAlive() {
        return alive;
    }
}

File: src/tasks/MainThreads.ts
import {ChromaNativeSDK} from '../razer-sdk/ChromaNativeSDK.js';
import {resetHttpClients} from '../rest/http/LeagueHttpClients.js';
import {RunningState} from '../state/RunningState.js';
import {fetchActivePlayerTask} from './FetchActivePlayerTask.js';
import {fetchGameStatsTask} from './FetchGameStatsTask.js';
import {fetchNewEventsTask} from './FetchNewEventsTask.js';
import {fetchPlayerListTask} from './FetchPlayerListTask.js';
import {fetchPlayerNameTask} from './FetchPlayerNameTask.js';
import {FixedDelayScheduler} from './FixedDelayScheduler.js';
import {GameLoader} from './GameLoader.js';
import {createRefreshMainHudTask} from './RefreshMainHudTask.js';

const PLAYER_LIST_FETCH_DELAY = 300;
const ACTIVE_PLAYER_FETCH_DELAY = 100;
const EVENTS_FETCH_DELAY = 300;
const MAIN_HUD_REFRESH_DELAY = 50;
const GAME_STATS_FETCH_DELAY = 300;

export class MainThreads {
    private readonly mainScheduler = new FixedDelayScheduler();
    private gameLoader: GameLoader | null = null;
    private readonly chromaNativeSDK: ChromaNativeSDK;
    private alive = true;

    constructor() {
        console.log('Loading game...');
        this.chromaNativeSDK = new ChromaNativeSDK();
    }

    async connect(): Promise<void> {
        await this.chromaNativeSDK.connect();
        // Start loading animation only after SDK is connected
        this.gameLoader = new GameLoader(this.chromaNativeSDK);
        this.mainScheduler.scheduleWithFixedDelay(fetchNewEventsTask, 0, EVENTS_FETCH_DELAY);
    }

    async initializeGameThreads(): Promise<void> {
        if (this.gameLoader) {
            this.gameLoader.close();
            this.gameLoader = null;
        }
        await fetchPlayerNameTask();
        console.log(`${RunningState.getGameState().playerRiotId} joined the game`);
        this.mainScheduler.scheduleWithFixedDelay(fetchGameStatsTask, 20, GAME_STATS_FETCH_DELAY);
        this.mainScheduler.scheduleWithFixedDelay(fetchPlayerListTask, 50, PLAYER_LIST_FETCH_DELAY);
        this.mainScheduler.scheduleWithFixedDelay(fetchActivePlayerTask, 50, ACTIVE_PLAYER_FETCH_DELAY);
        this.mainScheduler.scheduleWithFixedDelay(
            createRefreshMainHudTask(this.chromaNativeSDK), 150, MAIN_HUD_REFRESH_DELAY,
        );
    }

    async close(): Promise<void> {
        if (!this.alive) return;
        this.alive = false;
        RunningState.setRunningGame(false);
        this.gameLoader?.close();
        this.gameLoader = null;
        this.mainScheduler.shutdown();
        resetHttpClients();
        try {
            await this.chromaNativeSDK.close();
        } catch (ex) {
            console.warn('Exception while shutting down Chroma SDK', ex);
        }
        console.log('Player left the game, waiting for another...');
    }

    isAlive(): boolean {
        return this.alive;
    }
}
